src/backends/tensorrt/plan_backend.cc

```diff
@@ -3161,15 +3214,39 @@ PlanBackend::Context::Run(
             "failed to run TRT response");
       }
 
-      // Process the output tensors with the device memory address even if
-      // zero-copy may be used, so that the memory copies perform asynchronously
-      // and wait for model execution.
-      payload_->responder_->ProcessTensor(
-          name, dt, batchn_shape,
-          static_cast<const char*>(io_binding_info.device_buffer_),
-          TRITONSERVER_MEMORY_GPU, gpu_device_);
+      // Use pinned memory address if zero copy is supported, else use device
+      // memory address.
+      if (zero_copy_support_) {
+        LOG_ERROR << "cudaLaunchHostFunc ProcessTensor";
+        cudaStreamWaitEvent(stream_, events_[next_set_].ready_for_output_, 0);
+        payload_->process_tensor_tuples_[io_index] = {
+            name,
+            dt,
+            batchn_shape,
+            static_cast<const char*>(io_binding_info.buffer_),
+            io_binding_info.memory_type_,
+            io_binding_info.memory_type_id_,
+            &payload_->responder_};
+        cudaLaunchHostFunc(
+            stream_, ProcessTensorCudaHost,
+            reinterpret_cast<void*>(
+                &payload_->process_tensor_tuples_[io_index]));
+      } else {
+        payload_->responder_->ProcessTensor(
+            name, dt, batchn_shape,
+            static_cast<const char*>(io_binding_info.device_buffer_),
+            io_binding_info.memory_type_, io_binding_info.memory_type_id_);
+      }
     }
   }
+
+  if (zero_copy_support_) {
+    cudaStreamWaitEvent(stream_, events_[next_set_].output_ready_, 0);
+    LOG_ERROR << "cudaStreamWaitEvent for output_ready";
+    cudaStreamSynchronize(stream_);
```

Comment:
I don't think you should be syncing here.

Reply:
Fixed